Build error with aeson-2

When I build `shakespeare-2.0.25` with `aeson-2.0.1.0`, I get the following error:

```
Text/Julius.hs:120:23: error:
    • Couldn't match type ‘Data.Aeson.KeyMap.KeyMap Value’
                     with ‘H.HashMap T.Text Value’
      Expected type: H.HashMap T.Text Value
        Actual type: aeson-2.0.1.0:Data.Aeson.Types.Internal.Object
    • In the first argument of ‘H.toList’, namely ‘m’
      In the expression: H.toList m
      In the expression:
        case H.toList m of
          (x : xs) -> singleton '{' <> one x <> foldr f (singleton '}') xs
          _ -> "{}"
    |
120 |         case H.toList m of
    |                       ^
```

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -14,7 +14,12 @@ jobs:
       fail-fast: false
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
-        resolver: [nightly, lts-18, lts-16, lts-14]
+        stack_args:
+          - --resolver=nightly
+          - --resolver=lts-18
+          - --resolver=lts-16
+          - --resolver=lts-14
+          - --stack-yaml=stack-ghc-9.2.yaml
 
     steps:
       - name: Clone project
@@ -32,4 +37,4 @@ jobs:
               ls C:/ProgramData/Chocolatey/bin/
               rm C:/ProgramData/Chocolatey/bin/ghc*
             fi
-            stack test --fast --no-terminal --resolver=${{ matrix.resolver }}
+            stack test --fast --no-terminal ${{ matrix.stack_args }}

--- a/Text/Hamlet.hs
+++ b/Text/Hamlet.hs
@@ -139,7 +139,7 @@ bindingPattern (BindList is) = do
     return (ListP patterns, concat scopes)
 bindingPattern (BindConstr con is) = do
     (patterns, scopes) <- fmap unzip $ mapM bindingPattern is
-    return (ConP (mkConName con) patterns, concat scopes)
+    return (conP (mkConName con) patterns, concat scopes)
 bindingPattern (BindRecord con fields wild) = do
     let f (Ident field,b) =
            do (p,s) <- bindingPattern b
@@ -156,6 +156,13 @@ mkConName = mkName . conToStr
 conToStr :: DataConstr -> String
 conToStr (DCUnqualified (Ident x)) = x
 conToStr (DCQualified (Module xs) (Ident x)) = intercalate "." $ xs ++ [x]
+
+conP :: Name -> [Pat] -> Pat
+#if MIN_VERSION_template_haskell(2,18,0)
+conP name = ConP name []
+#else
+conP = ConP
+#endif
 
 -- Wildcards bind all of the unbound fields to variables whose name
 -- matches the field name.

--- a/Text/Julius.hs
+++ b/Text/Julius.hs
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE TemplateHaskell #-}
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
 {-# LANGUAGE FlexibleInstances #-}
@@ -53,6 +54,9 @@ import qualified Data.Text as TS
 import qualified Data.Text.Lazy as TL
 import Text.Shakespeare
 import Data.Aeson (Value, toJSON)
+#if MIN_VERSION_aeson(2,0,0)
+import qualified Data.Aeson.KeyMap as KeyMap
+#endif
 import Data.Aeson.Types (Value(..))
 import Numeric (showHex)
 import qualified Data.HashMap.Strict as H
@@ -117,11 +121,17 @@ encodeToTextBuilder =
                       V.foldr f (singleton ']') (V.unsafeTail v)
       where f a z = singleton ',' <> go a <> z
     go (Object m) = {-# SCC "go/Object" #-}
-        case H.toList m of
+        case fromObject m of
           (x:xs) -> singleton '{' <> one x <> foldr f (singleton '}') xs
           _      -> "{}"
       where f a z     = singleton ',' <> one a <> z
             one (k,v) = string k <> singleton ':' <> go v
+
+#if MIN_VERSION_aeson(2,0,0)
+    fromObject = H.toList . KeyMap.toHashMapText
+#else
+    fromObject = H.toList
+#endif
 
 string :: T.Text -> Builder
 string s = {-# SCC "string" #-} singleton '"' <> quote s <> singleton '"'

--- a/Text/MkSizeType.hs
+++ b/Text/MkSizeType.hs
@@ -38,7 +38,7 @@ showInstanceDec name unit' = instanceD [] (instanceType "Show" name) [showDec]
         x = mkName "x"
         unit = LitE $ StringL unit'
         showDec = FunD (mkName "show") [Clause [showPat] showBody []]
-        showPat = ConP name [VarP x]
+        showPat = conP name [VarP x]
         showBody = NormalB $ AppE (AppE showSize $ VarE x) unit
 
 numInstanceDec :: Name -> Dec
@@ -65,8 +65,8 @@ instanceType className name = AppT (ConT $ mkName className) (ConT name)
 
 binaryFunDec :: Name -> String -> Dec
 binaryFunDec name fun' = FunD fun [Clause [pat1, pat2] body []]
-  where pat1 = ConP name [VarP v1]
-        pat2 = ConP name [VarP v2]
+  where pat1 = conP name [VarP v1]
+        pat2 = conP name [VarP v2]
         body = NormalB $ AppE (ConE name) result
         result = AppE (AppE (VarE fun) (VarE v1)) (VarE v2)
         fun = mkName fun'
@@ -75,7 +75,7 @@ binaryFunDec name fun' = FunD fun [Clause [pat1, pat2] body []]
 
 unariFunDec1 :: Name -> String -> Dec
 unariFunDec1 name fun' = FunD fun [Clause [pat] body []]
-  where pat = ConP name [VarP v]
+  where pat = conP name [VarP v]
         body = NormalB $ AppE (ConE name) (AppE (VarE fun) (VarE v))
         fun = mkName fun'
         v = mkName "v"
@@ -92,3 +92,10 @@ notStrict = Bang NoSourceUnpackedness NoSourceStrictness
 
 instanceD :: Cxt -> Type -> [Dec] -> Dec
 instanceD = InstanceD Nothing
+
+conP :: Name -> [Pat] -> Pat
+#if MIN_VERSION_template_haskell(2,18,0)
+conP name = ConP name []
+#else
+conP = ConP
+#endif

--- a/Text/Shakespeare/I18N.hs
+++ b/Text/Shakespeare/I18N.hs
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE TemplateHaskell #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE MultiParamTypeClasses #-}
@@ -186,7 +187,7 @@ toClauses prefix dt (lang, defs) =
         (pat, bod) <- mkBody dt (prefix ++ constr def) (map fst $ vars def) (content def)
         guard <- fmap NormalG [|$(return $ VarE a) == pack $(lift $ unpack lang)|]
         return $ Clause
-            [WildP, ConP (mkName ":") [VarP a, WildP], pat]
+            [WildP, conP (mkName ":") [VarP a, WildP], pat]
             (GuardedB [(guard, bod)])
             []
 
@@ -230,7 +231,7 @@ sToClause :: String -> String -> SDef -> Q Clause
 sToClause prefix dt sdef = do
     (pat, bod) <- mkBody dt (prefix ++ sconstr sdef) (map fst $ svars sdef) (scontent sdef)
     return $ Clause
-        [WildP, ConP (mkName "[]") [], pat]
+        [WildP, conP (mkName "[]") [], pat]
         (NormalB bod)
         []
 
@@ -241,9 +242,16 @@ defClause = do
     d <- newName "msg"
     rm <- [|renderMessage|]
     return $ Clause
-        [VarP a, ConP (mkName ":") [WildP, VarP c], VarP d]
+        [VarP a, conP (mkName ":") [WildP, VarP c], VarP d]
         (NormalB $ rm `AppE` VarE a `AppE` VarE c `AppE` VarE d)
         []
+
+conP :: Name -> [Pat] -> Pat
+#if MIN_VERSION_template_haskell(2,18,0)
+conP name = ConP name []
+#else
+conP = ConP
+#endif
 
 toCon :: String -> SDef -> Con
 toCon dt (SDef c vs _) =

--- a/shakespeare.cabal
+++ b/shakespeare.cabal
@@ -45,7 +45,7 @@ library
                    , ghc-prim
                    , bytestring
                    , directory        >= 1.2
-                   , aeson
+                   , aeson            < 3
                    , blaze-markup
                    , blaze-html
                    , exceptions

--- a/stack-ghc-9.2.yaml
+++ b/stack-ghc-9.2.yaml
@@ -1,6 +1,10 @@
 resolver: nightly-2021-11-14
+compiler: ghc-9.2.1
 
 extra-deps:
   - aeson-2.0.2.0
+  - attoparsec-0.14.2
+  - base-compat-0.12.1
+  - base-compat-batteries-0.12.1
   # https://github.com/simonmar/happy/issues/215
   - happy-1.20.0

--- a/stack-ghc-9.2.yaml
+++ b/stack-ghc-9.2.yaml
@@ -1,0 +1,6 @@
+resolver: nightly-2021-11-14
+
+extra-deps:
+  - aeson-2.0.2.0
+  # https://github.com/simonmar/happy/issues/215
+  - happy-1.20.0

